python/cuml/neighbors/nearest_neighbors.pyx

```diff
@@ -291,8 +292,8 @@ class NearestNeighbors(Base):
 
         return m, expanded
 
-    def kneighbors(self, X=None, n_neighbors=None,
-                   return_distance=True, convert_dtype=True):
+    def kneighbors(self, X=None, n_neighbors=None, return_distance=True,
+                   convert_dtype=True, return_cupy=False):
```

Comment:
So I've tried implementing this and the idea makes sense but there's a few consequences I discovered and I wanted to hear your thoughts:

1. Making cumlarray an `output_type` for `NearestNeighbors` exposes it to the end user and based on what I read cumlarray is meant solely for internal cuML/RAPIDS developers.
2. When `kneighbors_graph` is called on `NearestNeighbors` I'll have to set `self.output_type='cumlarray'` and `self._mirror_input=False` prior to the internal `kneighbors` call and then revert them afterwards which doesn't feel great.
3. This looks like it may impact other models because it's stationed in `base.pyx`.

Other alternatives that I've thought of are having a `return_type` parameter for `kneighbors` if we want the generality of selecting the output type on the fly. Let me know how you think we should move forward on this one.

Reply:
@aleksficek, 

The output type for this method can override the setting of this variable when kneighbors is invoked with an `output_dtype` which is not `None`. That provides a feature to users while also making it possible for `kneighbors` to provide an output internally`. 

Let's pull the current implementation of `kneighbors` into an internal method `_kneighbors` and have `kneighbors()` just call it directly. That way we can add an additional option `_output_cumlarray` internally without having to expose it to the public API.

Reply:
Thanks for the response. I'm in the process of adding the `_output_cumlarray` parameter, should I also add the `output_type` variable you mentioned because I would just end up using `output_type='cupy'` instead of setting `_output_cumlarray`? On a side note, does `CumlArray` support list manipulation like `[:, 1:]` because I'm getting some CUDA runtime errors as a result?

Reply:
I don't think slicing would be too hard to add, as it'd just need to proxy to the device array underneath, but I don't think it's supported today. For now, I would do the slicing after calling `to_output`

Reply:
`output_type` should be on the `kneighbors_graph` convenience wrapper. It's already an option in the `NearestNeighbors` constructor so it should just be passed into there from the convenience wrapper.

Reply:
Awesome, that's what I have for `output_type`!. Just a heads up, anyone using `kneighbors` with `_output_cumlarray=True` && `X=None`, will have to post process the distances/indices array with a `[:,1:]` as well. This should be fine though because it will all be moved "under the hood" in https://github.com/rapidsai/cuml/issues/2562.